Makefile

```diff
@@ -263,10 +263,26 @@ clean-build-image:
 	rm -f .scBuildImage
 	docker rmi -f scbuildimage > /dev/null 2>&1 || true
 
+# clean-generated does a `git checkout --` on all generated files and
+# directories.  May not work correctly if you have staged some of these files
+# or have multiple commits.
 clean-generated:
 	rm -f .generate_files
+	# rollback changes to generated defaults/conversions/deepcopies
+	find $(TOP_SRC_DIRS) -name zz_generated* | xargs git checkout --
+	# rollback changes to types.generated.go
+	find $(TOP_SRC_DIRS) -name types.generated* | xargs git checkout --
+	# rollback changes to the generated clientset directories
+	find $(TOP_SRC_DIRS) -type d -name *_generated | xargs git checkout --
+	# rollback openapi changes
+	git checkout -- pkg/openapi/openapi_generated.go
+
+# purge-generated removes generated files from the filesystem.  Note:
+# types.generated.go is incredibly picky and not currently covered by this
+# target.
```

Comment:
for clarity, the echo command will populate `types.generated.go` with the minimum necessary to sucessfully run. Why the codec gen fails, I don't know. That's the one code generator that is external to kubernetes.